### PR TITLE
Room notifications for threads

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes https://github.com/vector-im/element-web/issues/19126

Blocked by https://github.com/matrix-org/matrix-js-sdk/pull/2007

Screen Shot 2021-10-29 at 11 13 13

---
This PR currently has no changelog labels, so will not be included in changelogs.

Add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is plus `X-Breaking-Change` if it's a breaking change.

Preview: https://619fc9b40d3ba6135df28213--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/stores/notifications/NotificationColor.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export enum NotificationColor {
     // TODO: Remove bold with notifications: https://github.com/vector-im/element-web/issues/14227
     Bold, // no badge, show as unread
     Grey, // unread notified messages
+    BoldRed, // no badge, show as unread and notify there is a mention
```
Comment (turt2live) [inline src/stores/notifications/NotificationColor.ts:23]: I'm not sure I fully understand this, and adding fields here has implications beyond notification states: the room list needs to consider this case.

File: src/stores/notifications/ThreadNotificationState.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import { NotificationColor } from "./NotificationColor";
+import { NotificationCountType } from "matrix-js-sdk/src/@types/receipt";
+import { NotificationState } from "./NotificationState";
+import { IDestroyable } from "../../utils/IDestroyable";
+import { Thread, ThreadEvent } from "matrix-js-sdk/src/models/thread";
+
+export class ThreadNotificationState extends NotificationState implements IDestroyable {
+    public _symbol: string = null;
+    public _count = 0;
```
Comment (turt2live) [inline src/stores/notifications/ThreadNotificationState.ts:25]: These shouldn't be public.

File: src/stores/notifications/RoomThreadsNotificationState.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+limitations under the License.
+*/
+
+import { NotificationColor } from "./NotificationColor";
+import { NotificationCountType } from "matrix-js-sdk/src/@types/receipt";
+import * as RoomNotifs from '../../RoomNotifs';
+import * as Unread from '../../Unread';
+import { RoomNotificationState } from "./RoomNotificationState";
+
+export class RoomThreadsNotificationState extends RoomNotificationState {
```
Comment (turt2live) [inline src/stores/notifications/RoomThreadsNotificationState.ts:23]: the amount of copy/paste in this class is uncomfortable - why can't `RoomNotificationState` be used?

**Review by turt2live — CHANGES_REQUESTED** [review]
The code overall seems fine, though I'm confused by the notification state usage/implementation here: any possible state for our badges should be represented by the colours we already have (at least until we refactor the notifications system), so we shouldn't need to add any new ones. 

A new state for threads themselves makes sense, but I am confused as to the need for a room thread state.

Is there design documentation which can accompany the notification changes here?

File: src/stores/notifications/NotificationColor.ts (hunk 1 above)
Comment (gsouquet) [reply src/stores/notifications/NotificationColor.ts:23]: The designs introduces a new state where a badge is displayed without the notification count but with a differentiator if a user has been mentioned or not 

Screen Shot 2021-11-01 at 11 29 53

The consideration I took when adding this was to place it in the correct order as the enum will be translated to an auto increment numeric value.
The code in `NotificationState.ts` is never checking hardcoded numerical value but relies more on the order of the `NotificationColor` enum properties.

`hasUnreadCount` will return `false`, and `hasMentions` will return `true`.

Let me know if you can think of any other cases that are not covered here

File: src/stores/notifications/RoomThreadsNotificationState.ts (hunk 3 above)
Comment (gsouquet) [reply src/stores/notifications/RoomThreadsNotificationState.ts:23]: There's a fair amount of duplicated code and I'll probably give a shot at DRYing it up

However there's some slight differences in how things are counted and cases that are relevant for the room level badge, but that aren't relevant for the threads.

**Comment by gsouquet** [comment]
> A new state for threads themselves makes sense, but I am confused as to the need for a room thread state.
Is there design documentation which can accompany the notification changes here?

I attempted to highlight that in the screenshot I originally posted in the PR description. 

You can find all the design explorations here https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=1246%3A231502 but let me explain a bit better what I'm attempting to do.

I identified the need to have 3 different types of badges
- `RoomNotificationState` The room list item badge, which displays the total number of unread messages in the room timeline and the sum of all unread messages in all the threads in that room 
- `RoomThreadsNotificationState` The room header thread icon badge, which displays the sum of all unread messages in all the threads in that room
- `ThreadNotificationState` The thread panel item badge, which displays a coloured dot alongside the event to notify if there are unread messages in that thread

It looked like the right approach to extend `NotificationState` to create these badges as this is what the `NotificationBadge` component consumes.

Me and the Thread's team understand that folks have different expectations regarding notifications and I fully expect to iterate on the information we surface to the end user

File: src/stores/notifications/NotificationColor.ts (hunk 1 above)
Comment (turt2live) [reply src/stores/notifications/NotificationColor.ts:23]: Right, that sort of state is meant to be handled by the UI code (ignore the count from the notification state) - we shouldn't need a new colour for that.

File: src/stores/notifications/ThreadNotificationState.ts [hunk 4]
```diff
@@ -21,8 +21,8 @@ import { IDestroyable } from "../../utils/IDestroyable";
 import { Thread, ThreadEvent } from "matrix-js-sdk/src/models/thread";
 
 export class ThreadNotificationState extends NotificationState implements IDestroyable {
-    public _symbol: string = null;
-    public _count = 0;
+    protected _symbol: string = null;
+    protected _count = 0;
```
Comment (turt2live) [inline src/stores/notifications/ThreadNotificationState.ts:25]: are these properties used? It doesn't appear so

**Comment by gsouquet** [comment]
The tests are passing locally. Branch matching seems to be broken again

**Review by turt2live — APPROVED** [review]
thanks! Sorry the notification state stuff is so complicated :(

**Comment by turt2live** [comment]
I think if you merge develop then it'll fix itself - looks like github's breaking change spanned this PR's lifetime

**Pull request closed without merging**